python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+# Python SDK - Phase 2
+
+The Python SDK will be implemented in three loosely defined phases:
+
+* Phase 1 - Initial project, activity worker, and client
+* Phase 2 - Workflow worker, local activities, async activity completion, etc
+* Phase 3 - Deterministic workflow sandbox, replayer, test suite (some parts may be combined with phase 2)
+
+This proposal is for phase 2. It is intentionally loosely defined and all aspects are subject to discussion/change.
+
+## Overview
+
+Following up from phase 1 and known implementation changes since then, this phase will implement the actual workflow
+worker/code and other advanced features.
+
+A simple example of a workflow definition and worker run:
+
+```python
+from datetime import timedelta
+from temporalio import activity, workflow
+from temporalio.client import Client
+from temporalio.worker import Worker
+
+@activity.defn
+async def say_hello(name: str) -> str:
+    return f"Hello, {name}!"
+
+@workflow.defn
+class MyWorkflow:
+    @workflow.run
+    async def run(arg1: str) -> str:
+        return await workflow.execute_local_activity(
+            say_hello, arg1, start_to_close_timeout=timedelta(seconds=5)
+        )
+
+
+async def main(stop_event: asyncio.Event):
+    client = await Client.connect("http://localhost:7233")
+    async with Worker(
+        client,
+        task_queue="somequeue",
+        activities=[say_hello],
+        workflows=[MyWorkflow],
+    ):
+        await stop_event.wait()
+```
+
+## Workflow Definitions
+
+A workflow is a class.
+
+Here is an example of a class-based workflow that calls a local activity to get
+a greeting with a signal to update the prefix and a query to get the latest
+greeting:
+
+```python
+import asyncio
+from datetime import timedelta
+from temporalio import activity, workflow
+from temporalio.client import Client
+from temporalio.worker import Worker
+
+@activity.defn
+async def create_greeting_activity(salutation: str, name: str) -> str:
+    return f"{salutation}, {name}!"
+
+@workflow.defn
```

Comment:
What is the value of having the `workflow.defn` tag? IF it does something - what? If not, why have it?

Reply:
It marks a class as a workflow and all the validation/constraints that come with that and it is a place where the name is set (or derived from the thing below it). We have no other options at this time, but it is important to use as a way to mark this class as a workflow with those inherent constraints.